ground/gcs/src/plugins/config/outputchannelform.cpp

```diff
@@ -386,15 +386,30 @@ void OutputChannelForm::updateMaxSpinboxValue(UAVObject *obj)
         foreach(qint32 channel, channelBank) {
             // ... and if there's a match, set the maximum values and return
             if (channel-1 == m_index) {
-                if (actuatorSettingsData.TimerUpdateFreq[i] == 0)
-                    return;
-
-                double maxPulseWidth = round(10000000.0 / actuatorSettingsData.TimerUpdateFreq[i]);
-
-                // Saturate at the UAVO's maximum value
-                if (maxPulseWidth > std::numeric_limits<__typeof__(actuatorSettingsData.ChannelMax[0])>::max())
-                    maxPulseWidth = std::numeric_limits<__typeof__(actuatorSettingsData.ChannelMax[0])>::max();
-
+                
+                double maxPulseWidth = 0;
+                
+                if (actuatorSettingsData.TimerUpdateFreq[i] != 0)
+                {
+                    double pwm_freq = 0;
+                    if (actuatorSettingsData.TimerPwmResolution[i] == ActuatorSettings::TIMERPWMRESOLUTION_1MHZ)
+                        pwm_freq = 1000000;
+                    else if (actuatorSettingsData.TimerPwmResolution[i] == ActuatorSettings::TIMERPWMRESOLUTION_12MHZ)
+                        pwm_freq = 12000000;
+                    maxPulseWidth = round(pwm_freq / actuatorSettingsData.TimerUpdateFreq[i]);
+                    
+                    // Saturate at the UAVO's maximum value
+                    if (maxPulseWidth > std::numeric_limits<__typeof__(actuatorSettingsData.ChannelMax[0])>::max())
+                        maxPulseWidth = std::numeric_limits<__typeof__(actuatorSettingsData.ChannelMax[0])>::max();
+                } else {
+                    // SyncPWM has been selected, the actual pulse limit is a function of sensor rate.
+                    // Simply limit to 1ms here to account for 1000Hz sensor rate.
+                    if (actuatorSettingsData.TimerPwmResolution[i] == ActuatorSettings::TIMERPWMRESOLUTION_1MHZ)
+                        maxPulseWidth = 1000;
+                    else if (actuatorSettingsData.TimerPwmResolution[i] == ActuatorSettings::TIMERPWMRESOLUTION_12MHZ)
+                        maxPulseWidth = 12000;
```

Comment:
Per Peabody's comments the pwm time should be limited to the max sensor
update rate. It was previously limited only to the size of the UAVO (16 bit
unsigned).

Yes it effectively makes sync-PWM and low resolution unusable. But this is
not a mode that is typically used by anybody. If desired, and if low enough
sensor rates are used, the user can set whatever pwm time he chooses with
the UAVOs directly.

A PWM count of 12000 at 12MHz is still 1 millisecond. It's my understanding
the values here are not in milliseconds but counts of 83.3 nanoseconds.

On Tue, Jun 2, 2015 at 2:40 PM, Michael Lyle <notifications@github.com>
wrote:

> In ground/gcs/src/plugins/config/outputchannelform.cpp
> <https://github.com/TauLabs/TauLabs/pull/1642#discussion_r31556025>:
>
> > +                    if (actuatorSettingsData.TimerPwmResolution[i] == ActuatorSettings::TIMERPWMRESOLUTION_1MHZ)
> > +                        pwm_freq = 1000000;
> > +                    else if (actuatorSettingsData.TimerPwmResolution[i] == ActuatorSettings::TIMERPWMRESOLUTION_12MHZ)
> > +                        pwm_freq = 12000000;
> > +                    maxPulseWidth = round(pwm_freq / actuatorSettingsData.TimerUpdateFreq[i]);
> > +
> > +                    // Saturate at the UAVO's maximum value
> > +                    if (maxPulseWidth > std::numeric_limits<__typeof__(actuatorSettingsData.ChannelMax[0])>::max())
> > +                        maxPulseWidth = std::numeric_limits<__typeof__(actuatorSettingsData.ChannelMax[0])>::max();
> > +                } else {
> > +                    // SyncPWM has been selected, the actual pulse limit is a function of sensor rate.
> > +                    // Simply limit to 1ms here to account for 1000Hz sensor rate.
> > +                    if (actuatorSettingsData.TimerPwmResolution[i] == ActuatorSettings::TIMERPWMRESOLUTION_1MHZ)
> > +                        maxPulseWidth = 1000;
> > +                    else if (actuatorSettingsData.TimerPwmResolution[i] == ActuatorSettings::TIMERPWMRESOLUTION_12MHZ)
> > +                        maxPulseWidth = 12000;
>
> The maximum pulse length has to be shorter than our maximum possible
> flight loop update rate, unless the underlying code correctly will inhibit
> a pulse if a pulse is already in progress. I've not read the underlying
> code to know.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/TauLabs/TauLabs/pull/1642/files#r31556025>.
>